pex/pex_builder.py

```diff
@@ -253,7 +256,48 @@ def _add_dist_dir(self, path, dist_name):
         self._copy_or_link(filename, target)
     return CacheHelper.dir_hash(path)
 
+  def _get_installer_paths(self, dist_name, base):
+    """Set up an overrides dict for WheelFile.install that installs the contents
+    of a wheel into its own base in the pex dependencies cache.
+    """
+    return {
+      'purelib': base,
+      'headers': os.path.join(base, 'headers'),
+      'scripts': os.path.join(base, 'bin'),
+      'platlib': base,
+      'data': base
+    }
+
   def _add_dist_zip(self, path, dist_name):
+    # We need to distinguish between wheels and other zips. Most of the time,
+    # when we have a zip, it contains its contents in an importable form.
+    # But wheels don't have to be importable, so we need to force them
+    # into an importable shape. We can do that by installing it into its own
+    # wheel dir.
+    if not self.interpreter.supports_wheel_install():
+      print("*** Wheel dependencies may not work correctly with Python 2.6.", file=sys.stderr)
+      print("*** Your generated pex may experience errors", file=sys.stderr)
+
+    if self.interpreter.supports_wheel_install() and dist_name.endswith("whl"):
+      from wheel.install import WheelFile
+      try:
+        tmp = tempfile.mkdtemp()
+        whltmp = os.path.join(tmp, dist_name)
+        os.mkdir(whltmp)
+        wf = WheelFile(path)
+        wf.install(overrides=self._get_installer_paths(dist_name, whltmp), force=True)
+        for (root, _, files) in os.walk(whltmp):
+          pruned_dir = os.path.relpath(root, tmp)
+          for f in files:
+            fullpath = os.path.join(root, f)
+            if os.path.isdir(fullpath):
+              continue
+            target = os.path.join(self._pex_info.internal_cache, pruned_dir, f)
+            self._chroot.copy(fullpath, target)
+      finally:
+        shutil.rmtree(tmp)
+      return CacheHelper.dir_hash(whltmp)
```

Comment:
won't the parent dir (`tmp`) of `whltmp` be removed before `dir_hash(whltmp)` is called?

assuming this was a bug and since this wasn't caught by existing tests, adding some extra test coverage here might be appropriate here if you're up for it.